File: p155.cpp
#include <ListNode.hpp>
#include <bits/stdc++.h>

using namespace std;

class MinStack {
private:
    int minimum;
    ListNode* head;
public:
    MinStack() {
        minimum = INT_MAX;
        head = NULL;
    }
    
    void push(int x) {
        if(x < this->minimum) {
            minimum = x;
        }
        ListNode* node = new ListNode(x);
        node->next = head;
        head = node;
    }
    
    void pop() {
        head = head->next;
        minimum = INT_MAX;
        ListNode* curr = head;
        while(curr) {
            if(curr->val < minimum) {
                minimum = curr->val;
            }
            curr = curr->next;
        }
    }
    
    int top() {
        return head->val;
    }
    
    int getMin() {
        return minimum;
    }
};

File: p155.py
class MinStack:

    def __init__(self):
        self.values = []
        self.mins = []

    def push(self, x: int) -> None:
        self.values.append(x)
        self.mins.append(x if len(self.mins) == 0 else min(x, self.mins[-1]))

    def pop(self) -> None:
        self.values.pop(-1)
        self.mins.pop(-1)

    def top(self) -> int:
        return self.values[-1]

    def getMin(self) -> int:
        return self.mins[-1]
